### PR TITLE
Albrja/mic 5383/add-number-steps-method

### PR DESCRIPTION
**Pull request by albrja**
### Adds method to SimulationContext to retreive number of steps
- *Category*: Feature
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5383

### Changes and notes
-adds method to InteractiveContext

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -1,3 +1,7 @@
+**3.0.10 - 10/07/24**
+
+  - Add method to InteractiveContext to get number of simulation steps
```
Comment (rmudambi) [inline CHANGELOG.rst:3]: This should read SimulationContext, not InteractiveContext

File: src/vivarium/framework/engine.py [hunk 2]
```diff
@@ -349,6 +350,26 @@ def get_population(self, untracked: bool = True) -> pd.DataFrame:
...
+            else None
+        )
+        if end_time and step_size:
+            num_steps = int(math.ceil((end_time - start_time) / step_size))
+        elif "simulation_events" in time_config:
+            num_steps = len(time_config["simulation_events"])
+        else:
+            ConfigurationKeyError(
+                "Time configuration is missing a necessary key. "
+                "'End' and 'step_size' or 'simulation_events' are required."
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:368]: Nit: `End` -> `end`

**Review by rmudambi — APPROVED** [review]
Let's go ahead with this for now to unblock the simulation model, but this solution isn't quite general enough to support all types of clocks, so we should update it later. 

Can you create another ticket to deal with that and put it in the backlog?

File: src/vivarium/framework/engine.py [hunk 3]
```diff
@@ -349,6 +350,26 @@ def get_population(self, untracked: bool = True) -> pd.DataFrame:
     def __repr__(self):
         return f"SimulationContext({self.name})"
 
+    def get_number_of_steps(self, start_time: pd.Timestamp) -> int:
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:353]: This method shouldn't take any arguments. It should be able to figure out the start time on its own.

**Review by rmudambi — CHANGES_REQUESTED** [review]

File: src/vivarium/framework/engine.py (hunk 2 above)
Comment (stevebachmeier) [inline src/vivarium/framework/engine.py:368]: nit: ambiguous. Maybe "Time configuration is incomplete. Either 'end' and 'step_size' values or a 'simulation_events' list are required."

File: src/vivarium/framework/engine.py [hunk 4]
```diff
@@ -349,6 +350,26 @@ def get_population(self, untracked: bool = True) -> pd.DataFrame:
...
+        time_config = self.configuration.time.to_dict()
+        end_time = pd.Timestamp(**time_config["end"]) if "end" in time_config else None
+        step_size = (
+            pd.Timedelta(days=time_config["step_size"])
+            if "step_size" in time_config
+            else None
+        )
+        if end_time and step_size:
+            num_steps = int(math.ceil((end_time - start_time) / step_size))
+        elif "simulation_events" in time_config:
```
Comment (stevebachmeier) [inline src/vivarium/framework/engine.py:363]: Should we handle the event where `simulation_events` AND end time or step size is provided (probably a raise?)?

**Review by stevebachmeier — COMMENTED** [review]
good to go once you sort out the start_time issue

File: tests/framework/test_time.py [hunk 5]
```diff
@@ -510,3 +510,42 @@ def test_step_size_post_processor(builder):
...
     assert np.all(value == pd.Timedelta(days=2))
+
+
+@pytest.mark.parametrize(
+    "end_day, clock",
+    [
+        (31, "DateTimeClock"),
+        (23, "DateTimeClock"),
+        (31, "SimpleClock"),
+        (23, "SimpleClock"),
```
Comment (rmudambi) [inline tests/framework/test_time.py:521]: The config of `SimpleClock` looks like this: 
```python
CONFIGURATION_DEFAULTS = {
        "time": {
            "start": 0,
            "end": 100,
            "step_size": 1,
            "standard_step_size": None,
        }
    }
```
Configuring it like this doesn't seem correct

**Review by rmudambi — APPROVED** [review]

**Review by stevebachmeier — APPROVED** [review]

**Pull request merged**